flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryArray.java

```diff
@@ -191,6 +191,22 @@ public Decimal getDecimal(int pos, int precision, int scale) {
 		return Decimal.readDecimalFieldFromSegments(segments, offset, offsetAndSize, precision, scale);
 	}
 
+	@Override
+	public SqlTimestamp getTimestamp(int pos, int precision) {
+		assertIndexIsValid(pos);
+
+		if (SqlTimestamp.isCompact(precision)) {
+			return SqlTimestamp.fromEpochMillis(segments[0].getLong(getElementOffset(pos, 8)));
```

Comment:
I think always reading from `segments[0]` is wrong

Reply:
Good catch, it should be `SegmentsUtil.getLong(segments, getElementOffset(pos, 8)`